### PR TITLE
Add `--iree-hal-list-target-backends` flag.

### PR DESCRIPTION
**Pull request by ScottTodd**
The HAL target backends are registered outside of the typical flags system, so `--help` output just appears as
```
--iree-hal-target-backends=<string>                          - Target backends for executable compilation.
```

This adds a new flag with sample usage:

```
λ iree-compile.exe --iree-hal-list-target-backends
Registered target backends:
  cuda
  llvm-cpu
  metal
  metal-spirv
  vmvx
  vmvx-inline
  vulkan
  vulkan-spirv
  webgpu
  webgpu-wgsl
```

Part of addressing https://github.com/iree-org/iree/issues/10432

---

I think we should also plumb something similar through to the Python bindings. This is what we have today:

```python
# Omit `target_backends`:
_ = compile_str(asm)

ValueError: Expected a non-empty list for 'target_backends'

# Try a backend that does not exist:
_ = compile_str(asm, target_backends=["unknown"])

CompilerToolError: Error invoking IREE compiler tool iree-compile
Diagnostics:
<stdin>:2:3: error: target backend 'unknown' not registered; registered backends: cuda, llvm-cpu, metal, metal, vmvx, vmvx-inline, vulkan, vulkan
  module @arithmetic {
  ^
compilation failed

# See a runtime helper method that we could model this after:
ireert.query_available_drivers()

['cuda', 'local-sync', 'local-task', 'vulkan']
```

### CODE REVIEW

**Review by benvanik — COMMENTED** [review]
I suspect there's a better way to do this but I can't immediately find it - maybe llvm::cl::callback that prints out the list and then calls exit(0)? this would prevent the need to pass in a source file or get a bunch of random garbage output. something like:
```
 cl::callback([](const bool &) {
   // do the printing here
   exit(0);
 }
```

looks like you can also add this (this is what llvm does for its own list flags):
`      cl::ValueDisallowed,`

**Comment by ScottTodd** [comment]
> I suspect there's a better way to do this but I can't immediately find it

Yeah, I'd like to find an elegant solution for this (especially if the list can just appear in the `--help` output, without needing to add a new flag). Same for `--iree-llvm-list-targets`. I figure that this might be good enough for now - I'd rather have an awkward way to get the information that no way at all, and I'm not sure how long it would take to figure out a better way (your ideas seem to be on the right track).

**Comment by benvanik** [comment]
I think at least giving the callback + exit a shot would be worth it - needing documentation on how to use a help flag is pretty self-defeating :P

File: compiler/src/iree/compiler/Tools/iree_compile_lib.cc [hunk 1]
```diff
@@ -161,6 +161,20 @@ int mlir::iree_compiler::runIreecMain(int argc, char **argv) {
       llvm::cl::desc("Split the input file into pieces and "
                      "process each chunk independently"),
       llvm::cl::init(false));
+  llvm::cl::opt<bool> listHalTargets(
+      "iree-hal-list-target-backends",
+      llvm::cl::desc(
+          "Lists all registered target backends for executable compilation."),
+      llvm::cl::init(false), llvm::cl::ValueDisallowed,
+      llvm::cl::callback([&](const bool &) {
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Tools/iree_compile_lib.cc:169]: Sweet, this works!

File: compiler/src/iree/compiler/Tools/iree_compile_lib.cc [hunk 2]
```diff
@@ -161,6 +161,20 @@ int mlir::iree_compiler::runIreecMain(int argc, char **argv) {
       llvm::cl::desc("Split the input file into pieces and "
                      "process each chunk independently"),
       llvm::cl::init(false));
+  llvm::cl::opt<bool> listHalTargets(
+      "iree-hal-list-target-backends",
+      llvm::cl::desc(
+          "Lists all registered target backends for executable compilation."),
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Tools/iree_compile_lib.cc:167]: I considered adding this flag to [`TargetRegistry.cpp`](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Dialect/HAL/Target/TargetRegistry.cpp) or [`TargetBackend.cpp`](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Dialect/HAL/Target/TargetBackend.cpp), but it didn't quite fit in, without a bit of refactoring. Is this location (`iree_compile_lib.cc`) okay?

**Review by benvanik — APPROVED** [review]

**Pull request merged**